### ISSUE TITLE
Double delete of vmdk when PV gets deleted.

### ISSUE THREAD

**Issue by mattmattox** [kubernetes/kubernetes#90770]
**What happened**:
When deleting a PV, the VMDK is deleted but the delete task is called twice. The first task successes but the second task fails because the VMDK is already deleted.

**What you expected to happen**:
Only one delete task should be created.

**How to reproduce it (as minimally and precisely as possible)**:
- Create PV using the default vsphere [storageclass](https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/storageclass.html) 
- Delete PV
- Watch logs in vCenter

**Anything else we need to know?**:
This issue is spamming the vCenter logs with the failed delete messages.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.15.5
- Cloud provider or hardware configuration: On-perm VMware
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04.04
- Kernel (e.g. `uname -a`): Unknown
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by leodotcloud** [comment]
I am not sure if the successive delete operations are triggered by k8s. The delete operations have the same opID with differing "sub" field. The same opID indicates that the second delete operation did not initiate from rancher/k8s but it could be some internal task triggered inside vCenter.


DELETE SUCCESS:
var/run/log/hostd.log:2020-06-19T00:08:31.300Z info hostd[2796274] [Originator@6876 sub=OsfsClient opID=15f91743-46-f3bc user=vpxuser:VSPHERE.LOCAL\someaccountname] Done deleting objects in directory: /vmfs/volumes/abcdefg-312344b4-1234-xyz628c87f90/kubevols/kubernetes-dynamic-pvc-12345abc-abcd-abcd-bfbb-abcd124axyzy.vmdk

DELETE FAILURE:
var/run/log/hostd.log:2020-06-19T00:08:31.303Z error hostd[2796274] [Originator@6876 sub=Nfcsvc opID=15f91743-46-f3bc user=vpxuser:VSPHERE.LOCAL\someaccountname] file delete force (/vmfs/volumes/abcdefg-312344b4-1234-xyz628c87f90/kubevols/kubernetes-dynamic-pvc-12345abc-abcd-abcd-bfbb-abcd124axyzy.vmdk -> /vmfs/volumes/5e14acac-3cfcc4b4-ea47-b02628c87f90/kubevols/kubernetes-dynamic-pvc-12345abc-abcd-abcd-bfbb-abcd124axyzy.vmdk) by 'vpxuser'

**Comment by leodotcloud** [comment]
[The error messages related to this issue were seen in a k8s 1.15.5 cluster]

**TL;DR:**
The second delete operation results in an error message in the kube-controller-manager logs, but it's harmless. Doesn't cause an inconsistent state.


**Detailed information:**

[PersistentVolumeController](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/controller/volume/persistentvolume/pv_controller_base.go#L277) handles the operations for PV and PVCs.

PVC add/update/delete operations are enqueued in `claimQueue` and later processed by [`claimWorker`](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/controller/volume/persistentvolume/pv_controller_base.go#L360)

Similarly, PV operations are enqueued in `volumeQueue` which are processed by [`volumeWorker`](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/controller/volume/persistentvolume/pv_controller_base.go#L302)

When a PVC is released, the corresponding PV is deleted (based on the reclaim policy) along with the backing VMDK file in vCenter. The delete operation for the PV is scheduled [here](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/controller/volume/persistentvolume/pv_controller.go#L1047), which ends up running in a separate goroutine [here](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/util/goroutinemap/goroutinemap.go#L108).

Since, the PV keys can be enqueued via multiple sources (informer, claimWorker, resync), the code to schedule the delete operation can get triggered multiple times. The majority of times, the key related to the delete operation is already found in the map and there is only a single delete operation.

But, there is a possibility when the key gets [deleted](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/util/goroutinemap/goroutinemap.go#L135) by a successful operation and the subsequent operation gets scheduled. This will result in the "double delete" messages as originally reported. The second delete operation will fail with "missing vmdk" from vCenter as the file was indeed deleted in the first delete operation.

The key for the PV gets enqueued back in the queue and by the time it gets processed again, the PV is deleted and the operation for the key completes.

There are no further retries for deleting the same PV as the PV doesn't exist anymore. This is quite evident from the logs collected as well.
